### ISSUE TITLE
Fine-Tuning Feedback Analysis

### ISSUE THREAD

**Issue by winstonallo** [winstonallo/42-stats#15]
_see #9 for start of this task_

There unfortunately is a gap in models fine-tuned for evaluating the constructiveness of feedback. This aspect is arguably the most important one in a peer-review setting, where the goal is not to classify sentiment, but to understand the constructive value of the feedback provided.
## Problem
Feedback analysis in pre-trained models is primarily focused on sentiment polarity (positive. negative, neutral), which is not (only) what we need here
## Proposed Solution
I propose fine-tuning of a model specifically for identifying and categorizing feedback as constructive or non-constructive. This could involve:
* **Dataset Creation:** Compiling a dataset of feedback examples rated by their constructiveness, either
  * specifically tuned to 42 feedbacks
  * or on a broader scale, maybe including PR reviews from GitHub
* **Training:** Finding a suitable model from the Transformers library (eg. BERT) and fine-tuning it to specifically recognize constructive feedback nuance
* **Evaluation Metrics:** Establishing clear metrics for evaluating the model's performance

**Comment by winstonallo** [comment]
I followed [this tutorial](https://medium.com/@coderhack.com/fine-tuning-bert-for-text-classification-a-step-by-step-guide-1a1c5f8e8ae1).
I am currently working from codespaces, which just terminated the process (because of excessive CPU use I suppose, the message only said `Terminated`)
I will have to test it on school computers when I'm back from vacation (I will also try on my laptop but idk if it can handle it).
If you want to try it yourself:
* checkout to [this branch](https://github.com/42-stats/42-stats/tree/15-fine-tuning-feedback-analysis)
* in **src/training/training_data** there are 2 csv files used for training (they have very small datasets, we will need to add more but I am focusing on getting it to work rn):
  * feedback_training.csv
  * feedback_test.csv
* The idea is to train the model using feedback comments labeled from 0 (not constructive at all) to 4 (very constructive)
* The _feedback_training.csv_ and _feedback_test.csv_ are both to be structured the same way: **text,label** _(intra feedback_,_rating you would give the feedback on scale mentioned above_)
* Run **src/training/feedback_model_training.py**, which consists of three classes:
  * **FeedbackDataset** (derived from torch.utils.data.Dataset), which sets custom rules for reading and tokenizing the csv
  * **BerttClassifier** (derived from torch.nn.Module), which combines the pre-trained BERT model with a linear classifier layer, used for fine-tuning on specific classification tasks
  * **Trainer**: there is a lot here (and I do not understand all of it), but basically, it uses the training csv to **train** the model, and then **evaluates** against expected outputs using the test csv

**Comment by winstonallo** [comment]
Sooo the training is working, we just need way more labeled data, I am using (very!!) small datasets for testing (32 lines lol),
so the model is being overtrained on details.
If you want to add data, follow the steps above, any help is welcome:)
You can also start your own new model and save it (caution with overwriting existing one!!)
